deploy/src/sidecar_state.rs

```diff
@@ -0,0 +1,195 @@
+        heartbeat_period: u64,
+        cluster_api: Api<Cluster>,
+        unreachable_thresh: usize,
+    ) -> Self {
+        Self {
+            statuses: HashMap::new(),
+            status_rx,
+            heartbeat_period,
+            cluster_api,
+            unreachable: HashMap::new(),
+            unreachable_thresh,
+        }
+    }
+
+    /// Task that update the state received from sidecar operators
+    pub(crate) async fn state_update_task(mut self) -> Result<()> {
+        let spec_size = self.get_spec_size().await.unwrap_or(2);
+        let majority = spec_size.overflow_add(1) / 2;
+        debug!("spec.size: {spec_size}, majority: {majority}");
+
+        loop {
+            while let Ok(status) = self.status_rx.recv_async().await {
+                debug!("reveived status: {status:?}");
+                let _prev = self.statuses.insert(status.id.clone(), status);
+
+                let Some(reachable_counts) =
+                    Self::get_reachable_counts(&self.statuses, self.heartbeat_period, majority)
+                else {
+                    continue;
+                };
+                debug!("reachable_counts: {reachable_counts:?}");
+
+                for id in self.statuses.keys() {
+                    let count = reachable_counts.get(id).copied().unwrap_or(0);
+                    // the sidecar operator is considered offline
+                    if count < majority {
+                        // if already in unreachable cache, increment the counter.
+                        // we would consider the recovery is failed if the counter reach
+                        // the threshold.
+                        if let Some(cnt) = self.unreachable.get_mut(id) {
+                            *cnt = cnt.overflow_add(1);
+                            if *cnt == self.unreachable_thresh {
+                                error!("failed to recover the operator: {id}");
+                                let _ignore = self.unreachable.remove(id);
+                                // TODO: notify the administrator
+                            }
+                        }
+                        // otherwise delete the pod, which will trigger k8s to recreate it
+                        else {
+                            debug!("{id} is unreachable, count: {count}, deleteing the pod");
+                            if let Err(e) =
+                                self.cluster_api.delete(id, &DeleteParams::default()).await
+                            {
+                                error!("failed to delete pod {id}, {e}");
+                            }
+                            let _ignore = self.unreachable.insert(id.clone(), 0);
+                        }
+                    // if recoverd, remove it from the cache
+                    } else if self.unreachable.remove(id).is_some() {
+                        debug!("operator {id} recoverd");
+                    } else {
+                    }
```

Comment:
The empty else here looks weird, could we print some debug information?

Reply:
done